### ISSUE TITLE
non-zero memory when shrinking after a calloc()

### ISSUE THREAD

**Issue by gemorin** [libhugetlbfs/libhugetlbfs#15]
There is some issue when some calloc() mem was not zeroed when using libhugetlbfs shrinking.

glibc by default expects "newly allocated" memory to be zeroed.  This is controlled by the MORECORE_CLEARS #define in the source code.  So it can't be modified without recompiling the library.

This is a problem when using shrinking since the code makes glibc believe that everything it asked for was trimmed while only full huge pages were really unmapped (and therefore will be zeroed when the heap grows again).

Since glibc only needs zeroing in some allocation paths, it would be wasteful imo to zero the memory that is "trimmed" but still mapped.  Therefore I think the best way is simply to tell glibc we only trimmed full pages.  To avoid that glibc tries to trim constantly an area we cannot unmap, I have upped the trim threshold.

Here is what I am suggesting:

```
--- a/morecore.c
+++ b/morecore.c
@@ -175,6 +175,7 @@ static void *hugetlbfs_morecore(ptrdiff_t increment)
                INFO("Attempting to unmap %ld bytes @ %p\n", -delta,
                        heapbase + mapsize + delta);
                ret = munmap(heapbase + mapsize + delta, -delta);
+                increment = heapbase - heaptop + mapsize + delta ;
                if (ret) {
                        WARNING("Unmapping failed while shrinking heap: "
                                "%s\n", strerror(errno));
@@ -357,7 +358,7 @@ void hugetlbfs_setup_morecore(void)
        /* Set some allocator options more appropriate for hugepages */

        if (__hugetlb_opts.shrink_ok)
-               mallopt(M_TRIM_THRESHOLD, hpage_size / 2);
+               mallopt(M_TRIM_THRESHOLD, hpage_size + hpage_size / 2);
        else
                mallopt(M_TRIM_THRESHOLD, -1);
        mallopt(M_TOP_PAD, hpage_size / 2);
```

**Comment by gemorin** [comment]
#17 for the pull request

**Comment by khers** [comment]
Please submit patches to the mailing list

**Comment by gemorin** [comment]
Also solved by 26c6b9b
